api.go

```diff
@@ -185,12 +189,108 @@ func (n notAllowedHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 }
 
 type resource struct {
-	resourceType reflect.Type
-	source       CRUD
-	name         string
-	marshalers   map[string]ContentMarshaler
+	resourceType   reflect.Type
+	source         CRUD
+	hooks          CRUDHooks
+	name           string
+	marshalers     map[string]ContentMarshaler
+
+	api            *API
+}
+
+func (res resource) buildRequest(r *http.Request, params httprouter.Params) (*Request, error) {
+	req := Request{
+		PlainRequest: r,
+		Header: r.Header,
+		Params: params,
+		QueryParams: r.URL.Query(),
+		Meta: make(map[string]interface{}),
+		Data: nil,
+	}
+
+	var body []byte
+
+	if r.Body != nil {
+		var err error
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			return nil, err
+		}
+		r.Body.Close()
+	}
+
+	if string(body) == "" {
+		req.Data = make(map[string]interface{})
+	} else {
+		data, err := unmarshalRequest(body, r, res.marshalers)
+		if err != nil {
+			return nil, errors.New("unmarshal_error: " + err.Error())
+		}
+		req.Data = data
+	}
+
+	if meta, ok := req.Data["meta"]; ok {
+		req.Meta = meta.(map[string]interface{})
```

Comment:
It works already, since the unmarshaler just returns the raw unmarshaled map[string]interface{}.
req.

I agree that meta should be removed from data when it exists.
Will do.